Audio: Volume behavior revamp, per-client-volume, fixes

kleinesfilmroellchen opened the pull request:
This is a collection of fixes/QOL/improvements to all parts of the audio subsystem:
- [x] Use a 0-1 volume instead of 0-100, which simplifies linear-log transformation logic and is more intuitive to program with
- [x] Logarithmic volume instead of the half-square-half-linear thing we had before
- [x] Fix deadlocks in AudioServer so that simultaneous audio playback actually works
- [x] Prevent clicks when changing volume quickly
- [x] Prevent clipping when mixing audio by reducing all client stream's volume (adding about 24dB of headroom)
- [x] Make Piano louder (old volume logic was dumb and written by me)
- [x] Each client can have its own volume (which is also logarithmic)

awesomekling reviewed (changes requested):
> I'm not sure why, but replacing the POSIX mutexes with Threading mutexes fixed this immediately.

You are not replacing POSIX mutexes (`Threading::Mutex` is a thin wrapper around a POSIX mutex).
What you are really doing is removing a condition variable.

Also it's mixed together with some unrelated class renaming changes.

Please make an effort to understand what you are changing, and improve patch hygiene by unbundling unrelated changes.

kleinesfilmroellchen commented:
I'm sorry @awesomekling , these commits were really hard to untangle (:yaktangle:). I'll try again.

I realize that I removed a condition variable, but as I could not understand why it was there in the first place, I initially removed it just to try. Then, I never encountered any deadlocks or other concurrency issues, even after testing (opening up to 7 audio applications, closing and reopening them etc.). So I concluded that the variable was not necessary. The mutex seemed logical from the start, and I just replaced it with the more readable Threading::Mutex. What I meant by "replacing POSIX mutexes" was that I removed the raw use of POSIX syscalls in favor of our library that (should) do things more correctly in general.

awesomekling commented:
cc @bugaevc

bugaevc commented on Userland/Libraries/LibIPC/Encoder.cpp:
```diff
@@ -106,6 +106,16 @@ Encoder& Encoder::operator<<(float value)
+        u64 as_u64;
+    } u;
+    u.as_double = value;
+    return *this << u.as_u64;
```
Isn't this [UB in C++](https://stackoverflow.com/questions/11373203/accessing-inactive-union-member-and-undefined-behavior)?

kleinesfilmroellchen replied:
It doesn't appear to be, as the float encoder does the same thing with u32 and I "copied" that. It also seems to me like that's what unions are for.

sunverwerth replied:
You could use `AK::bit_cast`

kleinesfilmroellchen replied:
Then I should rewrite the float code as well, if that's the better option.

bugaevc commented on Userland/Libraries/LibThreading/ConditionVariable.h:
```diff
@@ -0,0 +1,51 @@
+    ALWAYS_INLINE void wait_on(Mutex& to_wait_on)
+    {
+        pthread_cond_wait(&m_condition, &to_wait_on.m_mutex);
+    }
```
What would be really nice here is a method accepting a lambda to check the condition. Here's how it looks [in LWSP](https://github.com/bugaevc/lets-write-sync-primitives/blob/833be0d5e932c29a34d3dae8e01dccbbb0393570/src/condvar.cpp#L26):

```cpp
void CondVar::wait(std::function<bool()> condition) {
    while (!condition()) {
        wait();
    }
}
```

It's a small thing, but it makes actually using the condition variable so much cleaner and less error-prone.

bugaevc replied:
On that note, should `Threading::ConditionVariable` store a reference to its mutex, like LWSP condvar does, instead of requiring a mutex argument in `wait()` calls? I think that would be nicer.

kleinesfilmroellchen replied:
Yeah, I think I'll make it "connect" to a mutex "permanently".

I'm not sure what to take from your lambda suggestion. The wait() call doesn't return until it gets signalled, so why would you want a loop around it?

bugaevc replied:
No, `wait()` can return spuriously, that's just how it is. To quote POSIX:

> Spurious wakeups from the `pthread_cond_wait()` or `pthread_cond_timedwait()` functions may occur. Since the return from `pthread_cond_wait()` or `pthread_cond_timedwait()` does not imply anything about the value of this predicate, the predicate should be re-evaluated upon such return.

Moreover, even if `wait()` returns because of someone actually signaling the condition variable (and not spuriously), that doesn't yet mean the condition is now true: others may have changed it again before we grabbed the mutex.

So no, you always have to re-check the condition upon returning from `wait()`. And this is what I mean when I say wait-with-lambda makes this less error prone :slightly_smiling_face:

bugaevc commented on Userland/Services/AudioServer/Mixer.cpp:
```diff
@@ -65,12 +67,19 @@ void Mixer::mix()
+        if (m_pending_mixing.is_empty() && active_mix_queues.is_empty()) {
+            m_pending_mutex.lock();
+            m_mixing_necessary.wait_on(m_pending_mutex);
+            m_pending_mutex.unlock();
```
No no no, this is _not_ how you use condition variables!!

I'm missing some context about how AudioServer works; but you must first take the lock, then do the check, and then sleep, and then while still holding the lock act on the new state, and only then release the lock. And the modifications that you wait for must only happen with the same lock held. Otherwise you get races or deadlocks.

bugaevc replied:
Here's how this typically looks using that helper function I suggested earlier:

```cpp
// Worker
mutex.lock();
while (true) {
    condvar.wait([&] {
        return !queue.is_empty();
    });
    for (auto item : queue)
        process(item);
}

// Producer
mutex.lock();
queue.push(item);
mutex.unlock();
condvar.signal();
```

kleinesfilmroellchen replied:
The testing doesn't reveal such deadlocks. In turn, when the old code was still doing the queue modifications while the lock was held after returning from wait(), it deadlocked every time. So I'll try something like your suggestion, but I'm not sure if it will work.

(I haven't really written any synchronization code before so :shrug: )

kleinesfilmroellchen replied:
I've done it properly now, I think, but you'll have to take another look. As for the lambda thing, I'd still love an explanation why that would be useful in my case / in general.

bugaevc replied:
> I haven't really written any synchronization code before

Ah, let me actually expand on why you have to do it like this and not in a different way.

See, a **condition variable** is not just a way to wait until someone notifies you. It "oversees" some _condition_ that can _repeatedly_ become true or false during the run time of the program (if you only need to handle a condition that can change from false to true _once_, you're better off using a simpler "one-shot event" primitive).

The most typical use case (and the one we seem to have here as well) of such a condition is whether some queue (aka channel, socket, pipe) is not empty (contains something). You usually have a _producer_ (aka writer) and a _consumer_ (reader), and there can be multiple of both kinds. So without condition variables, with some simple "notification" primitive, you'd write something like this:

```c++
SomeQueue queue;
Notification queue_not_empty_notification;

void producer(SomeType item)
{
    queue.enqueue(item);
    queue_not_empty_notification.notify();
}

void consumer()
{
    while (true) {
        if (queue.is_empty())
            queue_not_empty_notification.wait();
        SomeType item = queue.dequeue();
        process(item);
    }
}
```

The obviously wrong this with this code is the lack of synchronization. It's Undefined Behavior to access the same data (the queue) for both reading and writing concurrently (non-atomically).

So let's add a mutex. Importantly, modifications _and_ checking of the condition must be done while holding the mutex (it's instant UB otherwise):

```c++
Mutex mutex;
SomeQueue queue;
Notification queue_not_empty_notification;

void producer(SomeType item)
{
    mutex.lock();
    queue.enqueue(item);
    mutex.unlock();

    queue_not_empty_notification.notify();
}

void consumer()
{
    mutex.lock();
    while (true) {
        if (queue.is_empty()) {
            // Release the mutex while waiting, to let
            // the producer(s) enqueue new items.
            mutex.unlock();
            queue_not_empty_notification.wait();
            mutex.lock();
        }
        SomeType item = queue.dequeue();
        process(item);
    }
}
```

Makes sense so far?

This code *almost* gets the job done. The remaining issue is _lost wakeups_. Namely, if a producer manages to grab the mutex, enqueue an item, release the mutex, and signal the notification — all **in between** the consumer releasing the mutex and starting to wait, then the wake-up signaled by the producer will be lost and the consumer would be left waiting forever for the notification that will never come.

So _that's_ the issue condition variables were designed to resolve. A condition variable works a lot like the "notification" primitive we've used above, except that it can magically do unlock-and-wait _atomically_. (Of course, there's no actual magic, just careful use of futexes.) This way no producer can sneak in between the unlock and wait calls: we atomically unlock the mutex and start waiting, there's no in-between state where we have already unlocked the mutex, but are not waiting yet.

This is what the code looks like with condition variables:

```c++
Mutex mutex;
SomeQueue queue;
CondVar queue_not_empty_condvar;

void producer(SomeType item)
{
    mutex.lock();
    queue.enqueue(item);
    mutex.unlock();

    queue_not_empty_condvar.notify();
}

void consumer()
{
    mutex.lock();
    while (true) {
        while (queue.is_empty())
            queue_not_empty_condvar.wait();
        SomeType item = queue.dequeue();
        process(item);
    }
}
```

The single `queue_not_empty_condvar.wait()` call replaces the three calls from the previous listing:

```c++
mutex.unlock();
queue_not_empty_notification.wait();
mutex.lock();
```

That is to say, `CondVar::wait()` unlocks the mutex, waits until notified, then locks the mutex back, so upon returning you own the mutex again. Unlike the notification primitive, it does unlock-then-wait atomically.

Condition variables also have another trick up their sleeve: they avoid "thundering hurd" in case of broadcast notifications. Suppose you have multiple worker/consumer threads (forming a thread pool), and you want them to process the items in parallel. Then you'd release the mutex while processing each item, to let other threads in:

```c++
Mutex mutex;
SomeQueue queue;
CondVar queue_not_empty_condvar;

void producer(SomeType item)
{
    mutex.lock();
    queue.enqueue(item);
    mutex.unlock();

    queue_not_empty_condvar.notify_all();
}

void consumer()
{
    while (true) {
        mutex.lock();
        while (queue.is_empty())
            queue_not_empty_condvar.wait();
        SomeType item = queue.dequeue();
        mutex.unlock();
        process(item);
    }
}
```

In a poorly written condition variable implementation, after the `notify_all()` call all the waiting worker threads would wake up and contend for the mutex; of course only one thread will get it; this is what's called the "thundering hurd" problem. In a properly written condition variable implementation, the `notify_all()` will only really wake one thread, but ensure that it will wake the next thread when it releases the mutex (once it dequeues an item), and so on, the threads will end up waking each other in a chain. But this is "just" an internal optimization, it doesn't affect the semantics of mutexes / condition variables.

To summarize:
* Condition variable is about a _condition_
* The condition must be modified and checked only when holding a specific mutex, to do otherwise is UB
* Compared to a simpler notification primitive, condition variables help you avoid lost wakeups by providing an atomic unlock-and-wait operation

I hope this clarifies things a bit :^)

kleinesfilmroellchen replied:
It makes sense now, also why the lambda is such a good idea. I recommend you make a blog post out of this text, it's too good of an explanation to be wasted on a ephermal github review comment :^). I'll have another shot and be back with a new version in a couple hours.

bugaevc commented on Userland/Services/AudioServer/Mixer.cpp:
```diff
@@ -65,12 +68,18 @@ void Mixer::mix()
-            pthread_mutex_lock(&m_pending_mutex);
-            pthread_cond_wait(&m_pending_cond, &m_pending_mutex);
+        // While we have nothing to mix, wait on the condition.
+        if (m_pending_mixing.is_empty() && active_mix_queues.is_empty()) {
```
This is the UB: you can't access `m_pending_mixing ` while not holding `m_pending_mutex `.

The pattern is, once again:

1. Lock the mutex.
2. Do the checks (while holding the mutex!).
3. If nothing to do (empty queue), wait on condition variable, then goto 2.
4. Act.
5. Release the mutex and go your merry way _OR_ (if in a loop) goto 2 without releasing the mutex.

In code:

```c++
mutex.lock();
condvar.wait([&] {
    return !queue.is_empty();
});
do_stuff(queue);
mutex.unlock();
```

Or better, don't hold the mutex while "doing stuff":

```c++
mutex.lock();
condvar.wait([&] {
    return !queue.is_empty();
});
auto stuff = dequeue_stuff(queue);
mutex.unlock();
do_stuff(stuff);
```

bugaevc commented on Userland/Libraries/LibThreading/ConditionVariable.h:
```diff
@@ -34,6 +36,11 @@ class ConditionVariable {
+    {
+        while (condition())
+            wait();
+    }
```
Shouldn't this go into the previous commit?

kleinesfilmroellchen replied:
Yes it should. 🤦‍♂️

bugaevc replied:
Also, the function in _Let's write synchronization primitives_ waits while the condition is _not_ true (so it's more of a "wait until" or "wait for" than "wait while"). I don't really have a preference either way, but keeping consistency with LWSP is a weak argument in favor of `wait_until()`.

What do y'all think?

bugaevc commented on Userland/Services/AudioServer/Mixer.cpp:
```diff
@@ -22,7 +23,8 @@ namespace AudioServer {
 
 Mixer::Mixer(NonnullRefPtr<Core::ConfigFile> config)
-    : m_device(Core::File::construct("/dev/audio", this))
+    : m_mixing_necessary(m_pending_mutex)
```
I wonder — can this be done at declaration site?

bugaevc commented on Userland/Services/AudioServer/Mixer.cpp:
```diff
@@ -65,13 +66,12 @@ void Mixer::mix()
-            pthread_cond_wait(&m_pending_cond, &m_pending_mutex);
+        m_pending_mutex.lock();
+        // While we have nothing to mix, wait on the condition.
+        m_mixing_necessary.wait_while([this, &active_mix_queues]() { return m_pending_mixing.is_empty() && active_mix_queues.is_empty(); });
```
I'm not too familiar with AudioServer internals, but can `active_mix_queues` possibly _not_ be empty here? Wouldn't we have handled that in the previous iteration?

And if I'm right about that, then the check below is redundant.

But other than that, yes, this looks correct 😀

kleinesfilmroellchen replied:
The reason we pause on active_mix_queues being empty is the whole point of this condition variable. The mixer thread will wait until a client connects, which will make the audio server not run while no audio has to be mixed and streamed. The check below is so that we only add a new queue if necessary. (There's now no condition around the entire lock thing, so we need to prevent unnecessary busywork). The condition will e.g. return immediately if we're mixing audio like normal, i.e. active_mix_queues has 1 or 2 clients and no client was added.

bugaevc replied:
Right, I see. The logic below doesn't drain the `active_mix_queues` list, it drains the queues themselves. So what you're doing here is if you already have active clients, you do a quick non-blocking check for new clients, and otherwise you wait. Good.

What would happen if you have some queues, but they are all empty? Is that possible?

kleinesfilmroellchen replied:
If a queue is empty, we don't append audio data for it (line 97). If that happens with all clients, we just don't change the current mixing buffer and stream all zeroes to the sound card. It's definitely possible, but I don't even know if it can happen with SoundPlayer while no file is loaded. That's a further optimization we could do, but it doesn't do harm because the few programs that do actually connect to the audio server usually want to stream audio.

bugaevc commented on Userland/Libraries/LibThreading/ConditionVariable.h:
```diff
@@ -0,0 +1,59 @@
+    // As with pthread APIs, the mutex must be locked or undefined behavior ensues.
+    ALWAYS_INLINE void wait()
+    {
+        pthread_cond_wait(&m_condition, &m_to_wait_on.m_mutex);
```
Technically I guess all these functions can return an error, let's `VERIFY()` that they don't.

bugaevc commented on Userland/Services/AudioServer/Mixer.cpp:
```diff
@@ -65,13 +66,12 @@ void Mixer::mix()
+        m_mixing_necessary.wait_while([this, &active_mix_queues]() { return m_pending_mixing.is_empty() && active_mix_queues.is_empty(); });
+        if (!m_pending_mixing.is_empty())
             active_mix_queues.extend(move(m_pending_mixing));
-            pthread_mutex_unlock(&m_pending_mutex);
```
Can we really rely on the `move()` clearing the vector? I'd rather we clear it explicitly after the move.

bugaevc reviewed (commented):
The condition variable-related parts LGTM, with two minor nitpicks

awesomekling commented on Userland/Services/AudioServer/Mixer.h:
```diff
@@ -125,9 +127,8 @@ class Mixer : public Core::Object {
-    pthread_mutex_t m_pending_mutex;
-    pthread_cond_t m_pending_cond;
+    Threading::Mutex m_pending_mutex;
+    Threading::ConditionVariable m_mixing_necessary = Threading::ConditionVariable(m_pending_mutex);
```
Please use brace initialization:
```cpp
Threading::ConditionVariable m_mixing_necessary { m_pending_mutex };
```
